File: backend/app/ingestion/lifecycle.py
"""Atomic replace + delete (P1 task #9 — FR-A2, FR-A3, NFR-4).

delete_document: purge a document and all its chunks (DB ON DELETE CASCADE
+ ORM cascade) so no stale content remains.

replace_document: ingest the new version into a NEW doc_id while the old stays
live; only when the new version is fully extracted/embedded does a SINGLE
transaction delete the old document and flip the new one to `ready`. If
anything fails, the old version stays live and the half-built new row is
discarded. There is no window where both versions are queryable, and none
where neither is.
"""

from __future__ import annotations

import uuid
from pathlib import Path

from sqlalchemy.orm import Session

from app.ingestion.guardrails import validate_path
from app.ingestion.pipeline import _populate
from app.models import Document


def delete_document(db: Session, doc_id: uuid.UUID) -> bool:
    """Delete a document and its chunks. Returns False if not found."""
    doc = db.get(Document, doc_id)
    if doc is None:
        return False
    db.delete(doc)  # cascade purges chunks (NFR-4: no stale content)
    db.commit()
    return True


def replace_document(
    db: Session,
    old_doc_id: uuid.UUID,
    *,
    file_path: str | Path,
    filename: str,
    category: str,
) -> Document:
    """Atomically replace `old_doc_id` with a freshly ingested new version.

    Raises ValueError if the old document is missing, GuardrailError if the
    new file violates a hard limit (old stays live in both cases).
    """
    old = db.get(Document, old_doc_id)
    if old is None:
        raise ValueError(f"document {old_doc_id} not found")

    # Hard guardrails before any work — old version stays live on rejection.
    num_pages = validate_path(file_path)

    new = Document(
        filename=filename,
        category=category,
        status="processing",  # invisible to queries until promoted
        page_count=num_pages,
        version=old.version + 1,
    )
    db.add(new)
    db.commit()
    db.refresh(new)

    # Build new version's chunks (staged, not committed).
    try:
        _populate(db, new, file_path)
    except Exception:
        db.rollback()
        db.delete(new)  # discard half-built new version
        db.commit()
        raise  # old untouched and still live

    # Atomic promote: delete old + commit new chunks + flip new->ready, one txn.
    try:
        db.delete(old)
        new.status = "ready"
        db.commit()
    except Exception:
        db.rollback()
        db.delete(new)
        db.commit()
        raise

    db.refresh(new)
    return new

File: mobile/app/src/lifecycle.ts
// Atomic replace + delete — port of backend/app/ingestion/lifecycle.py
// (FR-A2/A3, NFR-4).
//
// delete: remove a document; its chunks cascade (no stale content remains).
// replace: ingest the new version into a NEW id while the old stays live; only
// once the new version is fully extracted/embedded does a SINGLE transaction
// delete the old and flip the new to `ready`. On any failure the old version
// stays live and the half-built new row is discarded. There is no window where
// both versions are queryable, and none where neither is.
import { deleteDocumentRow, getDb, getDocument, insertDocument, setStatus } from "./db";
import { extractPdf } from "./extractClient";
import { checkFileSize, checkPageCount } from "./guardrails";
import { newId } from "./ids";
import {
  commitChunks,
  IngestInput,
  IngestResult,
  PreparedChunks,
  prepareChunks,
} from "./pipeline";

/** Delete a document and its chunks. Returns false if not found. */
export async function deleteDocument(id: string): Promise<boolean> {
  return deleteDocumentRow(id); // chunks cascade (NFR-4)
}

/** Atomically replace `oldId` with a freshly ingested new version. */
export async function replaceDocument(
  oldId: string,
  input: IngestInput,
): Promise<IngestResult> {
  checkFileSize(input.sizeBytes); // before any DB write — old stays live on reject

  const old = await getDocument(oldId);
  if (!old) throw new Error(`document ${oldId} not found`);

  const newDocId = newId();
  await insertDocument({
    id: newDocId,
    filename: input.filename,
    category: input.category,
    status: "processing", // invisible to queries until promoted
    page_count: 0,
    added_at: new Date().toISOString(),
    version: old.version + 1,
  });

  let prepared: PreparedChunks;
  let pageCount: number;
  try {
    const extract = await extractPdf(input.uri, input.filename);
    checkPageCount(extract.page_count);
    pageCount = extract.page_count;
    const db = await getDb();
    await db.runAsync(`UPDATE documents SET page_count = ? WHERE id = ?`, pageCount, newDocId);
    prepared = await prepareChunks(extract); // build new fully, old untouched
  } catch (e) {
    await deleteDocumentRow(newDocId); // discard half-built new; old still live
    throw e;
  }

  // Atomic promote: insert new chunks + delete old + flip new->ready, one txn.
  const db = await getDb();
  let chunkCount = 0;
  try {
    await db.withTransactionAsync(async () => {
      chunkCount = await commitChunks(db, newDocId, prepared);
      await deleteDocumentRow(oldId); // old chunks cascade
      await db.runAsync(`UPDATE documents SET status = 'ready' WHERE id = ?`, newDocId);
    });
  } catch (e) {
    await setStatus(newDocId, "failed");
    await deleteDocumentRow(newDocId);
    throw e; // old untouched and still live
  }

  return { id: newDocId, page_count: pageCount, chunk_count: chunkCount };
}
